openedx/core/djangoapps/theming/helpers.py

```diff
@@ -52,3 +71,341 @@ def get_themed_template_path(relative_path, default_path, **kwargs):
+
+    Example:
+        >> get_project_root_name()
+        'lms'
+        # from studio
+        >> get_project_root_name()
+        'cms'
+
+    Returns:
+        (str): component name of platform e.g lms, cms
+    """
+    root = Path(settings.PROJECT_ROOT)
+    if root.name == "":
+        root = root.parent
+    return root.name
+
+
+def get_base_theme_dir():
+    """
+    Return base directory that contains all the themes.
+
+    Example:
+        >> get_base_theme_dir()
+        '/edx/app/edxapp/edx-platform/themes'
+
+    Returns:
+         (Path): Base theme directory path
+    """
+    themes_dir = settings.COMPREHENSIVE_THEME_DIR
+    if not isinstance(themes_dir, basestring):
+        raise ImproperlyConfigured("COMPREHENSIVE_THEME_DIR must be a string.")
+    return Path(themes_dir)
+
+
+def is_comprehensive_theming_enabled():
+    """
+    Returns boolean indicating whether comprehensive theming functionality is enabled or disabled.
+    Example:
+        >> is_comprehensive_theming_enabled()
+        True
+
+    Returns:
+         (bool): True if comprehensive theming is enabled else False
+    """
+    return True if settings.COMPREHENSIVE_THEME_DIR else False
+
+
+def get_site_theme_cache_key(site):
+    """
+    Return cache key for the given site.
+
+    Example:
+        >> site = Site(domain='red-theme.org', name='Red Theme')
+        >> get_site_theme_cache_key(site)
+        'theming.site.red-theme.org'
+
+    Parameters:
+        site (django.contrib.sites.models.Site): site where key needs to generated
+    Returns:
+        (str): a key to be used as cache key
+    """
+    cache_key = "theming.site.{domain}".format(
+        domain=site.domain
+    )
+    return cache_key
+
+
+def is_valid_hostname(hostname):
+    """
+    Return boolean indicating if given hostname is valid or not
+
+    Example:
+        >> is_valid_hostname('red-theme.org')
+        True
+
+    Parameters:
+        hostname (str): hostname that needs to be tested.
+    Returns:
+        (bool): True if given hostname is valid else False
+
+    """
+    if len(hostname) > 255 or "." not in hostname:
+        return False
+    if hostname[-1] == ".":
+        hostname = hostname[:-1]  # strip exactly one dot from the right, if present
+    if ":" in hostname:
+        hostname = hostname.split(":")[0]  # strip port number if present
+
+    allowed = re.compile(r"(?!-)[A-Z\d-]{1,63}(?<!-)$", re.IGNORECASE)
+    return all(allowed.match(x) for x in hostname.split("."))
+
+
+def cache_site_theme_dir(site, theme_dir):
+    """
+    Cache site's theme directory.
+
+    Example:
+        >> site = Site(domain='red-theme.org', name='Red Theme')
+        >> cache_site_theme_dir(site, 'red-theme')
+
+    Parameters:
+        site (django.contrib.sites.models.Site): site for to cache
+        theme_dir (str): theme directory for the given site
+    """
+    cache.set(get_site_theme_cache_key(site), theme_dir, settings.FOOTER_CACHE_TIMEOUT)
```

Comment:
As mentioned above, we should break this dependency on the branding API and introduce theming-specific configuration.